early_stopping_notes (#1182)
mosaicml/composer: docs/source/notes/early_stopping.rst

|:octagonal_sign:| Early Stopping
=================================

> Early stopping and threshold stopping halt training based on set criteria. In Composer, they are callbacks can be passed to the Trainer.

Reviewer: ```suggestion
Early stopping and threshold stopping halt training based on set criteria. In Composer, this functionality is implemented as callbacks which can be configured and passed to the Trainer.
```

Early Stopping
--------------

The :class:`.EarlyStopper` callback stops training if a provided metric does not improve over a certain ``patience`` window of time.

.. testcode::

    import torch
    from composer import Trainer
    from composer.callbacks.early_stopper import EarlyStopper

    early_stopper = EarlyStopper(
        monitor='Accuracy',
        dataloader_label='train',
        patience='50ba',
        comp=torch.greater,
        min_delta=0.01,
    )

    trainer = Trainer(
        model=model,
        train_dataloader=train_dataloader,
        eval_dataloader=eval_dataloader,
        optimizers=optimizer,
        callbacks=[early_stopper],
        max_duration="1ep",
    )

In the above example, the ``'train'`` label means the callback is tracking the ``Accuracy`` metric for the train_dataloader. The default for the evaluation dataloader is ``eval``.

> We also set ``patience='50ba'`` and ``min_delta=0.01`` which means that every 50 batches, if the Accuracy does not exceed the best recorded Accuracy by ``0.01``, training is stopped. The `comp` argument indiciates that 'better' here means higher accuracy.

Reviewer 2: ```suggestion
We also set ``patience='50ba'`` and ``min_delta=0.01`` which means that every 50 batches, if the Accuracy does not exceed the best recorded Accuracy by ``0.01``, training is stopped. The ``comp`` argument indicates that 'better' here means higher accuracy.
```

> We also set ``patience='50ba'`` and ``min_delta=0.01`` which means that every 50 batches, if the Accuracy does not exceed the best recorded Accuracy by ``0.01``, training is stopped. The `comp` argument indiciates that 'better' here means higher accuracy.

Reviewer: [nit] might be helpful to mention here that `patience` can be specified in different time units.  Or just point to the time abstraction for details.

Each Evaluator object is marked with a ``label`` field for logging, and a ``metrics`` field that accepts a single metric, list of metrics. These can be provided to the callbacks above to indiciate which metric to monitor.

In the example below, the callback will monitor the `Accuracy` metric only for the dataloader marked `eval_dataset1`.

.. testcode::

    from composer import Trainer, Evaluator
    from torchmetrics.classification.accuracy import Accuracy
    from composer.callbacks.early_stopper import EarlyStopper

    evaluator1 = Evaluator(
        label='eval_dataset1',
        dataloader=eval_dataloader,
        metrics=Accuracy()
    )

    evaluator2 = Evaluator(
        label='eval_dataset2',
        dataloader=eval_dataloader2,
        metrics=Accuracy()
    )

    early_stopper = EarlyStopper(
        monitor='Accuracy',
        dataloader_label='eval_dataset1',
        patience=1
    )

    trainer = Trainer(
        model=model,
        train_dataloader=train_dataloader,
        eval_dataloader=[evaluator1, evaluator2],
        optimizers=optimizer,
        callbacks=[early_stopper],
        max_duration='1ep',
    )

.. note::

>     When using these callbacks with :class:`.Evaluator` objects, make sure that the ``dataloader_label`` and ``label`` field match the desired :class:`.Evaluator`.

Reviewer 2: ```suggestion

    When using these callbacks with :class:`.Evaluator` objects, make sure that the ``dataloader_label`` and ``label`` field match the desired :class:`.Evaluator`.
```

|:octagonal_sign:| Early Stopping
=================================

Early stopping and threshold stopping halt training based on set criteria. In Composer, they are both implemented as callbacks that are passed to the Trainer.

.. code:: python

    trainer = Trainer(
        ...,
        callbacks=[early_stopper],
    )

Early Stopping
--------------

The EarlyStopper callback tracks a particular training or evaluation metric and stops training if the metric does not improve within a given time interval.

The EarlyStopper callback takes several parameters.

* `monitor`: The name of the metric to track.

> * `dataloader_label`: The `dataloader_label` identifies which dataloader the metric belongs to. By default, the train dataloader is labeled `train`, and the evaluation dataloader is labeled `eval`, but these values can be changed.)

Reviewer 3: ```suggestion
* `dataloader_label`: The `dataloader_label` identifies which dataloader the metric belongs to. By default, the train dataloader is labeled `train`, and the evaluation dataloader is labeled `eval`, but these values can be changed.
```
